diff --git a/guides/subscriptions.md b/guides/subscriptions.md
index a30f97e1..a9dffc60 100644
--- a/guides/subscriptions.md
+++ b/guides/subscriptions.md
@@ -63,7 +63,7 @@ In your `MyAppWeb.Endpoint` module add:
 use Absinthe.Phoenix.Endpoint
 ```
 
-For your socket, different configurations are used in `MyAppWeb.UserSocket` depending on what version of
+For your socket, different configurations are used depending on what version of
 Phoenix you're using.
 
 #### Phoenix 1.3 and 1.4
diff --git a/lib/absinthe/phase/schema/validation/default_enum_value_present.ex b/lib/absinthe/phase/schema/validation/default_enum_value_present.ex
index daa47b04..b4761575 100644
--- a/lib/absinthe/phase/schema/validation/default_enum_value_present.ex
+++ b/lib/absinthe/phase/schema/validation/default_enum_value_present.ex
@@ -33,20 +33,18 @@ defmodule Absinthe.Phase.Schema.Validation.DefaultEnumValuePresent do
         values = Enum.map(enum.values, & &1.value)
         value_list = Enum.map(values, &"\n * #{inspect(&1)}")
 
-        default_valid? =
-          List.wrap(default_value)
-          |> Enum.all?(fn default -> default in values end)
-
-        if not default_valid? do
-          detail = %{
-            value_list: value_list,
-            type: type,
-            default_value: default_value
-          }
-
-          node |> put_error(error(node, detail))
-        else
-          node
+        case value_conforms_to_enum(node.type, default_value, values) do
+          {:error, value} ->
+            detail = %{
+              value_list: value_list,
+              type: type,
+              default_value: value
+            }
+
+            node |> put_error(error(node, detail))
+
+          {:ok, _} ->
+            node
         end
 
       _ ->
@@ -58,6 +56,25 @@ defmodule Absinthe.Phase.Schema.Validation.DefaultEnumValuePresent do
     node
   end
 
+  defp value_conforms_to_enum(%Blueprint.TypeReference.List{of_type: of_type}, value, enum_values)
+       when is_list(value) do
+    value
+    |> Enum.map(&value_conforms_to_enum(of_type, &1, enum_values))
+    |> Enum.find({:ok, value}, &match?({:error, _}, &1))
+  end
+
+  defp value_conforms_to_enum(%_{of_type: of_type}, value, enum_values) do
+    value_conforms_to_enum(of_type, value, enum_values)
+  end
+
+  defp value_conforms_to_enum(_, value, enum_values) do
+    if value in enum_values do
+      {:ok, value}
+    else
+      {:error, value}
+    end
+  end
+
   defp error(node, data) do
     %Absinthe.Phase.Error{
       message: explanation(data),
diff --git a/lib/absinthe/schema.ex b/lib/absinthe/schema.ex
index dfbe2607..f5892a61 100644
--- a/lib/absinthe/schema.ex
+++ b/lib/absinthe/schema.ex
@@ -106,7 +106,6 @@ defmodule Absinthe.Schema do
 
       @after_compile unquote(__MODULE__)
       @before_compile unquote(__MODULE__)
-      @prototype_schema Absinthe.Schema.Prototype
 
       @schema_provider Absinthe.Schema.Compiled
 
@@ -330,7 +329,7 @@ defmodule Absinthe.Schema do
       end
 
       def __absinthe_prototype_schema__() do
-        @prototype_schema
+        @schema_provider.__absinthe_prototype_schema__(__MODULE__)
       end
     end
   end
diff --git a/lib/absinthe/schema/compiled.ex b/lib/absinthe/schema/compiled.ex
index a6d7a8be..0ea67cdc 100644
--- a/lib/absinthe/schema/compiled.ex
+++ b/lib/absinthe/schema/compiled.ex
@@ -30,4 +30,8 @@ defmodule Absinthe.Schema.Compiled do
   def __absinthe_interface_implementors__(schema_mod) do
     Module.concat([schema_mod, Compiled]).__absinthe_interface_implementors__
   end
+
+  def __absinthe_prototype_schema__(schema_mod) do
+    Module.concat([schema_mod, Compiled]).__absinthe_prototype_schema__
+  end
 end
diff --git a/lib/absinthe/schema/persistent_term.ex b/lib/absinthe/schema/persistent_term.ex
index a592ee64..35276be7 100644
--- a/lib/absinthe/schema/persistent_term.ex
+++ b/lib/absinthe/schema/persistent_term.ex
@@ -85,6 +85,12 @@ if Code.ensure_loaded?(:persistent_term) do
       |> Map.fetch!(:__absinthe_interface_implementors__)
     end
 
+    def __absinthe_prototype_schema__(schema_mod) do
+      schema_mod
+      |> get()
+      |> Map.fetch!(:__absinthe_prototype_schema__)
+    end
+
     @dialyzer {:nowarn_function, [get: 1]}
     defp get(schema) do
       :persistent_term.get({__MODULE__, schema})
diff --git a/lib/absinthe/schema/provider.ex b/lib/absinthe/schema/provider.ex
index ef9a1b77..e77fb074 100644
--- a/lib/absinthe/schema/provider.ex
+++ b/lib/absinthe/schema/provider.ex
@@ -25,4 +25,6 @@ defmodule Absinthe.Schema.Provider do
   @callback __absinthe_directives__(schema_identifier) :: Absinthe.Type.Directive.t()
 
   @callback __absinthe_interface_implementors__(schema_identifier) :: term
+
+  @callback __absinthe_prototype_schema__(schema_identifier) :: schema_identifier
 end
diff --git a/lib/absinthe/subscription/supervisor.ex b/lib/absinthe/subscription/supervisor.ex
index 3f8d6d94..ce704692 100644
--- a/lib/absinthe/subscription/supervisor.ex
+++ b/lib/absinthe/subscription/supervisor.ex
@@ -4,15 +4,6 @@ defmodule Absinthe.Subscription.Supervisor do
   use Supervisor
 
   def start_link(pubsub, pool_size \\ System.schedulers_online() * 2) do
-    pubsub =
-      case pubsub do
-        [module] when is_atom(module) ->
-          module
-
-        module ->
-          module
-      end
-
     Supervisor.start_link(__MODULE__, {pubsub, pool_size})
   end
 
