diff --git a/lib/kino/bridge.ex b/lib/kino/bridge.ex
index 081e2d4..837f214 100644
--- a/lib/kino/bridge.ex
+++ b/lib/kino/bridge.ex
@@ -151,6 +151,16 @@ defmodule Kino.Bridge do
     end
   end
 
+  @doc """
+  Checks if the caller is running within Livebook context (group leader).
+  """
+  @spec within_livebook?() :: boolean()
+  def within_livebook?() do
+    # We make a Livebook-specific side-effect-free request and see if
+    # it is recognized
+    match?({:ok, _}, io_request(:livebook_get_evaluation_file))
+  end
+
   defp io_request(request) do
     gl = Process.group_leader()
     ref = Process.monitor(gl)
diff --git a/lib/kino/debug.ex b/lib/kino/debug.ex
index 6b342ec..77b2d52 100644
--- a/lib/kino/debug.ex
+++ b/lib/kino/debug.ex
@@ -11,15 +11,26 @@ defmodule Kino.Debug do
   falls back to the default backend otherwise.
   """
   @spec dbg(Macro.t(), Macro.t(), Macro.Env.t()) :: Macro.t()
-  def dbg(ast, _options, %Macro.Env{} = env) do
+  def dbg(ast, options, %Macro.Env{} = env) do
     dbg_id = System.unique_integer()
 
-    case ast do
-      {:|>, _meta, _args} ->
-        dbg_pipeline_ast(ast, dbg_id, env)
+    kino_ast =
+      case ast do
+        {:|>, _meta, _args} ->
+          dbg_pipeline_ast(ast, dbg_id, env)
 
-      _ ->
-        dbg_default_ast(ast, dbg_id, env)
+        _ ->
+          dbg_default_ast(ast, dbg_id, env)
+      end
+
+    fallback_ast = Macro.dbg(ast, options, env)
+
+    quote do
+      if Kino.Bridge.within_livebook?() do
+        unquote(kino_ast)
+      else
+        unquote(fallback_ast)
+      end
     end
   end
 
