diff --git a/lib/broadway.ex b/lib/broadway.ex
index 4c58769..0532bf2 100644
--- a/lib/broadway.ex
+++ b/lib/broadway.ex
@@ -297,13 +297,20 @@ defmodule Broadway do
       # Now assert the database side-effects
       ...
 
-  Keep in mind that multiple acknowledgement messages may be sent.
-  For example, if the batcher in the example above has size of 2,
-  then two batches would be created and therefore two ack messages
-  would be sent. Similarly, if any of the messages fail when
-  processed, an acknowledgement of their failure may be sent early
-  on. On the positive side, if you always push just a single test
-  message, then there is always one acknowledgment.
+  When testing pipelines that are using batchers there are additional
+  considerations. By default, the batch is only delivered when either
+  its size or its timeout has been reached, but that is often impractical
+  for testing, you may not necessarily want to send a lot of data or
+  wait a lot of time for the batch to flush. For this reason, when using
+  `test_messages/2`, by default the messages have their `:batch_mode`
+  set to `:flush` which means the batch will be immediately delivered.
+  Depending on the batch size and the batching mode you might get multiple
+  acknowledgment messages. For example, if the batcher in the example above
+  has size of 2, each event when batched would be immediately flushed and
+  so you would get a total of 3 acknowledment messages, one for each event.
+  Similarly, if any of the messages fail when processed, an acknowledgement
+  of their failure may be sent early on. On the positive side, if you always
+  push just a single test message, then there is always one acknowledgment.
   """
 
   alias Broadway.{BatchInfo, Message, Options, Server, Producer}
@@ -533,6 +540,12 @@ defmodule Broadway do
 
   See "Testing" section in module documentation for more information.
 
+  ## Options
+
+  * `:batch_mode` - when set to `:flush`, the batch the message is
+    in is immediately delivered. When set to `:bulk`, batch is
+    delivered when its size or timeout is reached. Defaults to `:flush`.
+
   ## Examples
 
   For example, in your tests, you may do:
@@ -544,10 +557,13 @@ defmodule Broadway do
 
   """
   @spec test_messages(GenServer.server(), data :: [term]) :: reference
-  def test_messages(broadway, data) when is_list(data) do
+  def test_messages(broadway, data, opts \\ []) when is_list(data) and is_list(opts) do
+    batch_mode = Keyword.get(opts, :batch_mode, :flush)
     ref = make_ref()
     ack = {Broadway.CallerAcknowledger, {self(), ref}, :ok}
-    messages = Enum.map(data, &%Message{data: &1, acknowledger: ack})
+
+    messages = Enum.map(data, &%Message{data: &1, acknowledger: ack, batch_mode: batch_mode})
+
     :ok = push_messages(broadway, messages)
     ref
   end
diff --git a/lib/broadway/batcher.ex b/lib/broadway/batcher.ex
index 311889b..0e80d73 100644
--- a/lib/broadway/batcher.ex
+++ b/lib/broadway/batcher.ex
@@ -81,29 +81,41 @@ defmodule Broadway.Batcher do
 
   defp handle_events_per_batch_key([%{batch_key: batch_key} | _] = events, acc, state) do
     {current, pending_count, timer} = init_or_get_batch(batch_key, state)
-    {current, pending_count, events} = split_counting(batch_key, events, pending_count, current)
 
-    acc = deliver_or_update_batch(batch_key, current, pending_count, timer, acc, state)
+    {current, pending_count, events, flush?} =
+      split_counting(batch_key, events, pending_count, false, current)
+
+    acc = deliver_or_update_batch(batch_key, current, pending_count, flush?, timer, acc, state)
     handle_events_per_batch_key(events, acc, state)
   end
 
-  defp split_counting(batch_key, [%{batch_key: batch_key} = event | events], count, acc)
-       when count > 0,
-       do: split_counting(batch_key, events, count - 1, [event | acc])
+  defp split_counting(batch_key, [%{batch_key: batch_key} = event | events], count, flush?, acc)
+       when count > 0 do
+    flush? = flush? or event.batch_mode == :flush
+    split_counting(batch_key, events, count - 1, flush?, [event | acc])
+  end
 
-  defp split_counting(_batch_key, events, count, acc), do: {acc, count, events}
+  defp split_counting(_batch_key, events, count, flush?, acc), do: {acc, count, events, flush?}
 
-  defp deliver_or_update_batch(batch_key, current, 0, timer, acc, state) do
-    delete_batch(batch_key)
-    cancel_batch_timeout(timer)
-    [wrap_for_delivery(batch_key, current, state) | acc]
+  defp deliver_or_update_batch(batch_key, current, _pending_count, true, timer, acc, state) do
+    deliver_batch(batch_key, current, timer, acc, state)
+  end
+
+  defp deliver_or_update_batch(batch_key, current, 0, _flush?, timer, acc, state) do
+    deliver_batch(batch_key, current, timer, acc, state)
   end
 
-  defp deliver_or_update_batch(batch_key, current, pending_count, timer, acc, _state) do
+  defp deliver_or_update_batch(batch_key, current, pending_count, _flush?, timer, acc, _state) do
     put_batch(batch_key, {current, pending_count, timer})
     acc
   end
 
+  defp deliver_batch(batch_key, current, timer, acc, state) do
+    delete_batch(batch_key)
+    cancel_batch_timeout(timer)
+    [wrap_for_delivery(batch_key, current, state) | acc]
+  end
+
   ## General batch handling
 
   defp init_or_get_batch(batch_key, state) do
diff --git a/lib/broadway/message.ex b/lib/broadway/message.ex
index f311ae9..c60b194 100644
--- a/lib/broadway/message.ex
+++ b/lib/broadway/message.ex
@@ -20,6 +20,7 @@ defmodule Broadway.Message do
           acknowledger: {module, ack_ref :: term, data :: term},
           batcher: atom,
           batch_key: term,
+          batch_mode: :bulk | :flush,
           status: :ok | {:failed, reason :: binary}
         }
 
@@ -29,6 +30,7 @@ defmodule Broadway.Message do
             acknowledger: nil,
             batcher: :default,
             batch_key: :default,
+            batch_mode: :bulk,
             status: :ok
 
   @doc """
@@ -61,6 +63,23 @@ defmodule Broadway.Message do
     %Message{message | batch_key: batch_key}
   end
 
+  @doc """
+  Sets the batching mode for the message.
+
+  When the mode is `:bulk`, the batch that the message is in is delivered after
+  the batch size or the batch timeout is reached.
+
+  When the mode is `:flush`, the batch that the message is in is delivered
+  immediately after processing. Note it doesn't mean the batch contains only a single element
+  but rather that all messages receives from the processor are delivered without waiting.
+
+  The default mode for messages is `:bulk`.
+  """
+  @spec put_batch_mode(message :: Message.t(), mode :: :bulk | :flush) :: Message.t()
+  def put_batch_mode(%Message{} = message, mode) when mode in [:bulk, :flush] do
+    %Message{message | batch_mode: mode}
+  end
+
   @doc """
   Mark a message as failed.
 
