diff --git a/lib/tailwind_formatter.ex b/lib/tailwind_formatter.ex
index 8591d23..2fedb89 100644
--- a/lib/tailwind_formatter.ex
+++ b/lib/tailwind_formatter.ex
@@ -93,7 +93,7 @@ defmodule TailwindFormatter do
       dynamic_classes
       |> Enum.reduce(contents, fn
         {dynamic_class, index}, contents ->
-          String.replace(contents, dynamic_class, "$#{index}")
+          String.replace(contents, dynamic_class, "$#{index}$")
       end)
 
     {:ok, placeholder_contents, dynamic_classes}
@@ -103,7 +103,7 @@ defmodule TailwindFormatter do
     dynamic_classes
     |> Enum.reduce(sorted_html, fn
       {dynamic_class, index}, html ->
-        String.replace(html, "$#{index}", dynamic_class)
+        String.replace(html, "$#{index}$", dynamic_class)
     end)
   end
 
