diff --git a/lib/absinthe/subscription.ex b/lib/absinthe/subscription.ex
index e6bddec2..e4386019 100644
--- a/lib/absinthe/subscription.ex
+++ b/lib/absinthe/subscription.ex
@@ -176,7 +176,7 @@ defmodule Absinthe.Subscription do
     registry = pubsub |> registry_name
 
     for field_key <- pdict_fields(doc_id) do
-      Registry.unregister(registry, field_key)
+      Registry.unregister_match(registry, field_key, doc_id)
     end
 
     Registry.unregister(registry, doc_id)
