diff --git a/lib/ash/error/exception.ex b/lib/ash/error/exception.ex
index 1aca65b2..b73dbbb3 100644
--- a/lib/ash/error/exception.ex
+++ b/lib/ash/error/exception.ex
@@ -55,7 +55,11 @@ defmodule Ash.Error.Exception do
             }
           end)
 
-        super(opts)
+        super(opts) |> Map.update(:vars, [], &clean_vars/1)
+      end
+
+      defp clean_vars(vars) do
+        vars |> Kernel.||([]) |> Keyword.drop([:field, :message, :path])
       end
 
       defoverridable exception: 1, message: 1
diff --git a/lib/ash/type/keyword.ex b/lib/ash/type/keyword.ex
index 476e4669..d840dbc2 100644
--- a/lib/ash/type/keyword.ex
+++ b/lib/ash/type/keyword.ex
@@ -175,7 +175,7 @@ defmodule Ash.Type.Keyword do
             {:ok, [{field, field_value} | result]}
 
           {:error, errors} ->
-            {:error, Enum.map(errors, fn error -> [field: field, message: error[:message]] end)}
+            {:error, Enum.map(errors, fn error -> Keyword.put(error, :field, field) end)}
         end
 
       :error ->
diff --git a/lib/ash/type/map.ex b/lib/ash/type/map.ex
index 0fd05cb0..3672981b 100644
--- a/lib/ash/type/map.ex
+++ b/lib/ash/type/map.ex
@@ -154,7 +154,7 @@ defmodule Ash.Type.Map do
             {:ok, Map.put(result, field, field_value)}
 
           {:error, errors} ->
-            {:error, Enum.map(errors, fn error -> [field: field, message: error[:message]] end)}
+            {:error, Enum.map(errors, fn error -> Keyword.put(error, :field, field) end)}
         end
 
       :error ->
