diff --git a/lib/ash.ex b/lib/ash.ex
index 573ec2aa..9e0c52b0 100644
--- a/lib/ash.ex
+++ b/lib/ash.ex
@@ -1238,20 +1238,7 @@ defmodule Ash do
 
   #{Spark.Options.docs(@can_question_mark_opts)}
   """
-  @spec can?(
-          query_or_changeset_or_action ::
-            Ash.Query.t()
-            | Ash.Changeset.t()
-            | Ash.ActionInput.t()
-            | {Ash.Resource.t(), atom | Ash.Resource.Actions.action()}
-            | {Ash.Resource.t(), atom | Ash.Resource.Actions.action(), input :: map}
-            | {Ash.Resource.record(), atom | Ash.Resource.Actions.action()}
-            | {Ash.Resource.record(), atom | Ash.Resource.Actions.action(), input :: map},
-          actor :: term,
-          opts :: Keyword.t()
-        ) ::
-          boolean | no_return
-
+  @spec can?(Ash.Can.subject(), actor(), Keyword.t()) :: boolean() | no_return()
   @doc spark_opts: [{2, @can_question_mark_opts}]
   def can?(action_or_query_or_changeset, actor, opts \\ []) do
     domain = Ash.Helpers.domain!(action_or_query_or_changeset, opts)
@@ -1306,18 +1293,7 @@ defmodule Ash do
 
   #{Spark.Options.docs(@can_opts)}
   """
-  @spec can(
-          action_or_query_or_changeset ::
-            Ash.Query.t()
-            | Ash.Changeset.t()
-            | Ash.ActionInput.t()
-            | {Ash.Resource.t(), atom | Ash.Resource.Actions.action()}
-            | {Ash.Resource.t(), atom | Ash.Resource.Actions.action(), input :: map}
-            | {Ash.Resource.record(), atom | Ash.Resource.Actions.action()}
-            | {Ash.Resource.record(), atom | Ash.Resource.Actions.action(), input :: map},
-          actor :: term,
-          opts :: Keyword.t()
-        ) ::
+  @spec can(Ash.Can.subject(), actor(), Keyword.t()) ::
           {:ok, boolean | :maybe}
           | {:ok, true, Ash.Changeset.t() | Ash.Query.t()}
           | {:ok, true, Ash.Changeset.t(), Ash.Query.t()}
@@ -1731,6 +1707,7 @@ defmodule Ash do
   end
 
   @type record_or_records :: Ash.Resource.record() | [Ash.Resource.record()]
+  @type actor :: any()
 
   @doc """
   Load fields or relationships on already fetched records. See `load/3` for more information.
diff --git a/lib/ash/actions/create/bulk.ex b/lib/ash/actions/create/bulk.ex
index d7961e53..3bfaa745 100644
--- a/lib/ash/actions/create/bulk.ex
+++ b/lib/ash/actions/create/bulk.ex
@@ -531,7 +531,7 @@ defmodule Ash.Actions.Create.Bulk do
             {new_errors, new_error_count} =
               Process.delete({:bulk_create_errors, tmp_ref}) || {[], 0}
 
-            store_error(ref, new_errors, new_error_count)
+            store_error(ref, new_errors, opts, new_error_count)
 
             notifications = Process.get({:bulk_update_notifications, tmp_ref}) || []
             store_notification(ref, notifications, opts)
@@ -958,9 +958,10 @@ defmodule Ash.Actions.Create.Bulk do
     end)
   end
 
-  defp store_error(_ref, empty, _opts) when empty in [[], nil], do: :ok
+  defp store_error(ref, empty, opts, error_count \\ nil)
+  defp store_error(_ref, empty, _opts, _error_count) when empty in [[], nil], do: :ok
 
-  defp store_error(ref, error, opts) do
+  defp store_error(ref, error, opts, error_count) do
     if opts[:stop_on_error?] && !opts[:return_stream?] do
       throw({:error, Ash.Error.to_error_class(error), 0})
     else
@@ -974,11 +975,11 @@ defmodule Ash.Actions.Create.Bulk do
 
         Process.put(
           {:bulk_create_errors, ref},
-          {new_errors ++ errors, count + Enum.count(new_errors)}
+          {new_errors ++ errors, count + (error_count || Enum.count(new_errors))}
         )
       else
         {errors, count} = Process.get({:bulk_create_errors, ref}) || {[], 0}
-        Process.put({:bulk_create_errors, ref}, {errors, count + 1})
+        Process.put({:bulk_create_errors, ref}, {errors, count + (error_count || 1)})
       end
     end
   end
diff --git a/lib/ash/can.ex b/lib/ash/can.ex
index b5f2a91f..8bca8b59 100644
--- a/lib/ash/can.ex
+++ b/lib/ash/can.ex
@@ -1,8 +1,28 @@
 defmodule Ash.Can do
-  @moduledoc false
+  @moduledoc """
+  Contains the Ash.can function logic.
+  """
 
   require Ash.Query
 
+  @type subject ::
+          Ash.Query.t()
+          | Ash.Changeset.t()
+          | Ash.ActionInput.t()
+          | {Ash.Resource.t(), atom | Ash.Resource.Actions.action()}
+          | {Ash.Resource.t(), atom | Ash.Resource.Actions.action(), input :: map}
+          | {Ash.Resource.record(), atom | Ash.Resource.Actions.action()}
+          | {Ash.Resource.record(), atom | Ash.Resource.Actions.action(), input :: map}
+
+  @doc """
+  Returns whether an actor can perform an action, query, or changeset.
+
+  You should prefer to use `Ash.can?/3` over this module, directly.
+
+  Can raise an exception if return_forbidden_error is truthy in opts or there's an error.
+  """
+  @spec can?(subject(), Ash.Domain.t(), Ash.Resource.record(), Keyword.t()) ::
+          boolean() | no_return()
   def can?(action_or_query_or_changeset, domain, actor, opts \\ []) do
     opts =
       opts
@@ -31,6 +51,20 @@ defmodule Ash.Can do
     end
   end
 
+  @doc """
+  Returns a an ok tuple if the actor can perform the action, query, or changeset,
+  an error tuple if an error happens, and a ok tuple with maybe if maybe is set to true
+  or not set.
+
+  You should prefer to use `Ash.can/3` over this module, directly.
+
+  Note: `is_maybe` is set to `true`, if not set.
+  """
+  @spec can(subject(), Ash.Domain.t(), Ash.Resource.record(), Keyword.t()) ::
+          {:ok, boolean() | :maybe}
+          | {:ok, boolean(), term()}
+          | {:ok, boolean(), Ash.Changeset.t(), Ash.Query.t()}
+          | {:error, Ash.Error.t()}
   def can(action_or_query_or_changeset, domain, actor, opts \\ []) do
     opts = Keyword.put_new(opts, :maybe_is, :maybe)
     opts = Keyword.put_new(opts, :run_queries?, true)
diff --git a/lib/ash/changeset/changeset.ex b/lib/ash/changeset/changeset.ex
index db83ab40..6e53d9bb 100644
--- a/lib/ash/changeset/changeset.ex
+++ b/lib/ash/changeset/changeset.ex
@@ -1094,50 +1094,58 @@ defmodule Ash.Changeset do
   end
 
   @doc false
+  # Returns either an appropriate expression for an atomic condition or a value
+  # indicated that the condition cannot be handled atomically.
+  #
+  # Validation logic matches on failure. So, for example, `present(:field)` is
+  # going to _match_ when `:field` is `nil`. However, when applying this logic
+  # to a `where` condition, the opposite is desired. The end result is kinda
+  # ugly because it can end up reading like "not is not equal to" but
+  # ultimately produces the correct results.
+  @spec atomic_condition([{module(), keyword()}], Ash.Changeset.t(), map()) ::
+          {:atomic, Ash.Expr.t() | boolean()} | {:not_atomic, String.t()}
   def atomic_condition(where, changeset, context) do
     Enum.reduce_while(where, {:atomic, true}, fn {module, validation_opts},
-                                                 {:atomic, condition} ->
+                                                 {:atomic, condition_expr} ->
       case module.atomic(
              changeset,
              validation_opts,
              struct(Ash.Resource.Validation.Context, context)
            ) do
         :ok ->
-          {:cont, {:atomic, condition}}
-
-        {:error, _} ->
-          {:cont, {:atomic, false}}
-
-        [{:atomic, _, expr, _as_error} | rest] ->
-          exprs = [expr | Enum.map(rest, &elem(&1, 2))]
-
-          new_expr =
-            Enum.reduce(exprs, condition, fn expr, condition ->
-              if condition == true do
-                expr
-              else
-                expr(^condition and ^expr)
-              end
-            end)
-
-          {:cont, {:atomic, new_expr}}
+          {:cont, {:atomic, condition_expr}}
 
         {:atomic, _, expr, _as_error} ->
-          new_expr =
-            if condition == true do
-              expr
-            else
-              expr(^condition and ^expr)
-            end
+          {:cont, {:atomic, atomic_condition_expr(condition_expr, expr)}}
 
-          {:cont, {:atomic, new_expr}}
-
-        {:not_atomic, reason} ->
-          {:halt, {:not_atomic, reason}}
+        {:error, _} ->
+          # Error from the validator, so the validations should just fail with
+          # a `false` expression.
+          {:halt, {:atomic, false}}
+
+        {:not_atomic, _reason} = not_atomic ->
+          {:halt, not_atomic}
+
+        atomic_conditions when is_list(atomic_conditions) ->
+          atomic_conditions
+          |> Stream.map(fn {:atomic, _, expr, _as_error} -> expr end)
+          |> Enum.reduce(condition_expr, &atomic_condition_expr(&2, &1))
+          |> then(&{:cont, {:atomic, &1}})
       end
     end)
   end
 
+  # This is not expressly necessary as `expr(true and not ^new_expr)` would also
+  # work just fine, but the final output from omitting `true` is much easier to
+  # read if debugging.
+  defp atomic_condition_expr(true, expr) do
+    expr(not (^expr))
+  end
+
+  defp atomic_condition_expr(condition_expr, expr) do
+    expr(^condition_expr and not (^expr))
+  end
+
   defp atomic_params(changeset, action, params, opts) do
     if opts[:assume_casted?] do
       Enum.reduce_while(params, changeset, fn {key, value}, changeset ->
@@ -4609,6 +4617,10 @@ defmodule Ash.Changeset do
   end
 
   def manage_relationship(changeset, relationship, input, opts) do
+    if opts == [] do
+      IO.warn("Calling `manage_relationship` without any options will not do anything")
+    end
+
     opts =
       if opts[:type] == :replace do
         Logger.warning(
diff --git a/lib/ash/filter/filter.ex b/lib/ash/filter/filter.ex
index c01d82ee..98c1aae8 100644
--- a/lib/ash/filter/filter.ex
+++ b/lib/ash/filter/filter.ex
@@ -987,6 +987,7 @@ defmodule Ash.Filter do
             }
           })
           |> Ash.Query.select([])
+          |> Map.put(:filter, nil)
           |> Ash.can(actor,
             run_queries?: false,
             pre_flight?: false,
diff --git a/lib/ash/query/query.ex b/lib/ash/query/query.ex
index b54e25c4..1c101b92 100644
--- a/lib/ash/query/query.ex
+++ b/lib/ash/query/query.ex
@@ -688,14 +688,14 @@ defmodule Ash.Query do
   end
 
   defp cast_params(query, action, args, opts) do
-    skip_unknown_inputs = opts[:skip_unknown_inputs] || []
+    skip_unknown_inputs = opts[:skip_unknown_inputs] || action.skip_unknown_inputs
 
     Enum.reduce(args, query, fn {name, value}, query ->
       cond do
         has_argument?(action, name) ->
           set_argument(query, name, value)
 
-        :* in List.wrap(opts[:skip_unknown_inputs]) ->
+        :* in skip_unknown_inputs ->
           query
 
         name in skip_unknown_inputs ->
diff --git a/lib/ash/resource/validation/attribute_does_not_equal.ex b/lib/ash/resource/validation/attribute_does_not_equal.ex
index 4d58f794..f15c363d 100644
--- a/lib/ash/resource/validation/attribute_does_not_equal.ex
+++ b/lib/ash/resource/validation/attribute_does_not_equal.ex
@@ -41,7 +41,7 @@ defmodule Ash.Resource.Validation.AttributeDoesNotEqual do
   def validate(changeset, opts, _context) do
     value = Ash.Changeset.get_attribute(changeset, opts[:attribute])
 
-    if value == opts[:value] do
+    if Comp.equal?(value, opts[:value]) do
       {:error,
        [field: opts[:attribute], value: value]
        |> with_description(opts)
diff --git a/lib/ash/resource/validation/attribute_equals.ex b/lib/ash/resource/validation/attribute_equals.ex
index 56260ef9..6beb5eb4 100644
--- a/lib/ash/resource/validation/attribute_equals.ex
+++ b/lib/ash/resource/validation/attribute_equals.ex
@@ -53,7 +53,7 @@ defmodule Ash.Resource.Validation.AttributeEquals do
   def validate(changeset, opts, _context) do
     value = Ash.Changeset.get_attribute(changeset, opts[:attribute])
 
-    if value != opts[:value] do
+    if Comp.not_equal?(value, opts[:value]) do
       {:error,
        [field: opts[:attribute], value: value]
        |> with_description(opts)
