diff --git a/lib/norm/core/schema.ex b/lib/norm/core/schema.ex
index 972fec6..7a89c2b 100644
--- a/lib/norm/core/schema.ex
+++ b/lib/norm/core/schema.ex
@@ -30,6 +30,11 @@ defmodule Norm.Core.Schema do
     |> Enum.at(0)
   end
 
+  def key_present?(schema, key) do
+    schema.specs
+    |> Enum.any?(fn {name, _} -> name == key end)
+  end
+
   defimpl Norm.Conformer.Conformable do
     alias Norm.Conformer
     alias Norm.Conformer.Conformable
diff --git a/lib/norm/core/selection.ex b/lib/norm/core/selection.ex
index f4d2abc..39217df 100644
--- a/lib/norm/core/selection.ex
+++ b/lib/norm/core/selection.ex
@@ -52,9 +52,9 @@ defmodule Norm.Core.Selection do
   defp validate_selectors!(other), do: raise ArgumentError, "select expects a list of keys but received: #{inspect other}"
 
   defp assert_spec!(%Schema{}=schema, key) do
-    case Schema.spec(schema, key) do
-      nil -> raise SpecError, {:selection, key, schema}
-      spec -> spec
+    case Schema.key_present?(schema, key) do
+      false -> raise SpecError, {:selection, key, schema}
+      true -> Schema.spec(schema, key)
     end
   end
   defp assert_spec!(%__MODULE__{}, _key) do
