diff --git a/lib/cachex.ex b/lib/cachex.ex
index b2da970..46c86ce 100644
--- a/lib/cachex.ex
+++ b/lib/cachex.ex
@@ -1405,6 +1405,8 @@ defmodule Cachex do
     do: raise ExecutionError, message: Errors.long_form(value)
   defp unwrap_unsafe({ :error, value }) when is_binary(value),
     do: raise ExecutionError, message: value
+  defp unwrap_unsafe({ :error, %ExecutionError { stack: stack } = e }),
+    do: reraise e, stack
   defp unwrap_unsafe({ _state, value }),
     do: value
 end
diff --git a/lib/cachex/execution_error.ex b/lib/cachex/execution_error.ex
index 55ae5ca..f1b57e6 100644
--- a/lib/cachex/execution_error.ex
+++ b/lib/cachex/execution_error.ex
@@ -14,5 +14,5 @@ defmodule Cachex.ExecutionError do
   The default error message should always be overridden with a long
   error formas displayed inside `Cachex.Errors`.
   """
-  defexception message: "Error during action execution"
+  defexception message: "Error during action execution", stack: nil
 end
diff --git a/lib/cachex/services/courier.ex b/lib/cachex/services/courier.ex
index 50c88d6..d00cc51 100644
--- a/lib/cachex/services/courier.ex
+++ b/lib/cachex/services/courier.ex
@@ -19,6 +19,7 @@ defmodule Cachex.Services.Courier do
 
   # add some aliases
   alias Cachex.Actions.Put
+  alias Cachex.ExecutionError
 
   ##############
   # Public API #
@@ -40,7 +41,7 @@ defmodule Cachex.Services.Courier do
   """
   @spec dispatch(Spec.cache, any, (() -> any)) :: any
   def dispatch(cache() = cache, key, task) when is_function(task, 0),
-    do: service_call(cache, :courier, { :dispatch, key, task })
+    do: service_call(cache, :courier, { :dispatch, key, task, local_stack() })
 
   ####################
   # Server Callbacks #
@@ -64,7 +65,7 @@ defmodule Cachex.Services.Courier do
   # Due to the nature of the async behaviour, this call will return before
   # the task has been completed, and the :notify callback will receive the
   # results from the task after completion (regardless of outcome).
-  def handle_call({ :dispatch, key, task }, caller, { cache, tasks }) do
+  def handle_call({ :dispatch, key, task, stack }, caller, { cache, tasks }) do
     references =
       case Map.get(tasks, key, []) do
         [] ->
@@ -74,7 +75,13 @@ defmodule Cachex.Services.Courier do
               try do
                 task.()
               rescue
-                e -> { :error, Exception.message(e) }
+                e -> {
+                  :error,
+                  %ExecutionError {
+                    message: Exception.message(e),
+                    stack: stack_compat() ++ stack
+                  }
+                }
               end
 
             normalized = normalize_commit(result)
@@ -105,4 +112,17 @@ defmodule Cachex.Services.Courier do
 
     { :noreply, { cache, Map.delete(tasks, key) } }
   end
+
+  ###############
+  # Private API #
+  ###############
+
+  # Generates a stack trace prior to dispatch.
+  defp local_stack do
+    self()
+    |> Process.info(:current_stacktrace)
+    |> elem(1)
+    |> tl
+    |> tl
+  end
 end
diff --git a/lib/cachex/spec.ex b/lib/cachex/spec.ex
index 6067ce7..7f086f4 100644
--- a/lib/cachex/spec.ex
+++ b/lib/cachex/spec.ex
@@ -464,6 +464,18 @@ defmodule Cachex.Spec do
   defmacro via(action, options),
     do: quote(do: [ { :via, unquote(action) } | unquote(options) ])
 
+  @doc """
+  Retrieves the currently handled stacktrace.
+  """
+  @spec stack_compat :: any
+  defmacro stack_compat() do
+    if Version.match?(System.version(), ">= 1.7.0") do
+      quote(do: __STACKTRACE__)
+    else
+      quote(do: System.stacktrace())
+    end
+  end
+
   @doc """
   Wraps a value inside a tagged Tuple using the provided tag.
   """
