diff --git a/lib/curl_req.ex b/lib/curl_req.ex
index 4099c5c..568b0c3 100644
--- a/lib/curl_req.ex
+++ b/lib/curl_req.ex
@@ -142,30 +142,23 @@ defmodule CurlReq do
 
   ## Examples
 
-      iex> CurlReq.from_curl("curl https://www.example.com")
-      %Req.Request{
-        method: :get, 
-        url: URI.parse("https://www.example.com"), 
-        registered_options: MapSet.new([:connect_options]), 
-        options: %{connect_options: [protocols: [:http1, :http2]]}
-      }
-
-      iex> CurlReq.from_curl("curl -I https://example.com")
-      %Req.Request{
-        method: :head, 
-        url: URI.parse("https://example.com"), 
-        registered_options: MapSet.new([:connect_options]), 
-        options: %{connect_options: [protocols: [:http1, :http2]]}
-      }
-
-      iex> CurlReq.from_curl("curl -b cookie_key=cookie_val https://example.com")
-      %Req.Request{
-        method: :get, 
-        headers: %{"cookie" => ["cookie_key=cookie_val"]}, 
-        url: URI.parse("https://example.com"), 
-        registered_options: MapSet.new([:connect_options]), 
-        options: %{connect_options: [protocols: [:http1, :http2]]}
-      }
+      iex> req = CurlReq.from_curl("curl https://www.example.com")
+      iex> req.method
+      :get
+      iex> req.url
+      URI.parse("https://www.example.com")
+
+      iex> req = CurlReq.from_curl("curl -I https://example.com")
+      iex> req.method
+      :head
+      iex> req.url
+      URI.parse("https://example.com")
+
+      iex> req = CurlReq.from_curl("curl -b cookie_key=cookie_val https://example.com")
+      iex> req.method
+      :get
+      iex> Req.Request.get_header(req, "cookie")
+      ["cookie_key=cookie_val"]
   """
   @doc since: "0.98.4"
 
@@ -188,30 +181,23 @@ defmodule CurlReq do
 
   ## Examples
 
-      iex> ~CURL(curl "https://www.example.com")
-      %Req.Request{
-        method: :get, 
-        url: URI.parse("https://www.example.com"), 
-        registered_options: MapSet.new([:connect_options]), 
-        options: %{connect_options: [protocols: [:http1, :http2]]}
-      }
-
-      iex> ~CURL(curl -I "https://example.com")
-      %Req.Request{
-        method: :head, 
-        url: URI.parse("https://example.com"), 
-        registered_options: MapSet.new([:connect_options]), 
-        options: %{connect_options: [protocols: [:http1, :http2]]}
-      }
-
-      iex> ~CURL(curl -b "cookie_key=cookie_val" "https://example.com")
-      %Req.Request{
-        method: :get, 
-        headers: %{"cookie" => ["cookie_key=cookie_val"]}, 
-        url: URI.parse("https://example.com"), 
-        registered_options: MapSet.new([:connect_options]), 
-        options: %{connect_options: [protocols: [:http1, :http2]]}
-      }
+      iex> req = ~CURL(curl "https://www.example.com")
+      iex> req.method
+      :get
+      iex> req.url
+      URI.parse("https://www.example.com")
+
+      iex> req = ~CURL(curl -I "https://example.com")
+      iex> req.method
+      :head
+      iex> req.url
+      URI.parse("https://example.com")
+
+      iex> req = ~CURL(curl -b "cookie_key=cookie_val" "https://example.com")
+      iex> req.method
+      :get
+      iex> Req.Request.get_header(req, "cookie")
+      ["cookie_key=cookie_val"]
   """
   defmacro sigil_CURL(curl_command, modifiers)
 
diff --git a/lib/curl_req/curl.ex b/lib/curl_req/curl.ex
index 0f7345f..8cfd276 100644
--- a/lib/curl_req/curl.ex
+++ b/lib/curl_req/curl.ex
@@ -145,6 +145,8 @@ defmodule CurlReq.Curl do
     %CurlReq.Request{}
     |> CurlReq.Request.put_url(url)
     |> CurlReq.Request.put_encoding(:form)
+    |> CurlReq.Request.put_compression(false)
+    |> CurlReq.Request.put_redirect(false)
     |> add_header(options)
     |> add_body(options)
     |> add_cookie(options)
diff --git a/lib/curl_req/req.ex b/lib/curl_req/req.ex
index 3988438..67b447e 100644
--- a/lib/curl_req/req.ex
+++ b/lib/curl_req/req.ex
@@ -89,7 +89,7 @@ defmodule CurlReq.Req do
   @spec encode(CurlReq.Request.t()) :: Req.Request.t()
   def encode(%CurlReq.Request{} = request, _opts \\ []) do
     req =
-      %Req.Request{}
+      Req.new()
       |> Req.merge(url: request.url)
       |> Req.merge(method: request.method)
 
@@ -138,6 +138,9 @@ defmodule CurlReq.Req do
       case request.compression do
         :none ->
           req
+          |> Req.Request.register_options([:compressed])
+          |> Req.Request.prepend_request_steps(compressed: &Req.Steps.compressed/1)
+          |> Req.merge(compressed: false)
 
         _ ->
           req
@@ -150,6 +153,9 @@ defmodule CurlReq.Req do
       case request.redirect do
         false ->
           req
+          |> Req.Request.register_options([:redirect])
+          |> Req.Request.prepend_response_steps(redirect: &Req.Steps.redirect/1)
+          |> Req.merge(redirect: false)
 
         _ ->
           req
diff --git a/mix.exs b/mix.exs
index 64922f0..cafdb9a 100644
--- a/mix.exs
+++ b/mix.exs
@@ -11,7 +11,8 @@ defmodule CurlReq.MixProject do
       package: package(),
       source_url: "https://github.com/derekkraan/curl_req",
       start_permanent: Mix.env() == :prod,
-      version: "0.100.0"
+      version: "0.100.0",
+      elixirc_paths: elixirc_paths(Mix.env())
     ]
   end
 
@@ -49,4 +50,7 @@ defmodule CurlReq.MixProject do
       maintainers: ["Derek Kraan", "Kevin Schweikert"]
     ]
   end
+
+  defp elixirc_paths(:test), do: ["lib", "test"]
+  defp elixirc_paths(_), do: ["lib"]
 end
